File: Day #1251/Solution.java
// Regex matching with '.' and '*' via DP; dp[i][j] = s[i:] matches p[j:]. O(m*n) time and space.
public class Solution {
    static boolean isMatch(String s, String p) {
        int m = s.length(), n = p.length();
        boolean[][] dp = new boolean[m + 1][n + 1];
        dp[m][n] = true;
        for (int i = m; i >= 0; --i) {
            for (int j = n - 1; j >= 0; --j) {
                boolean first = i < m && (p.charAt(j) == s.charAt(i) || p.charAt(j) == '.');
                if (j + 1 < n && p.charAt(j + 1) == '*') {
                    dp[i][j] = dp[i][j + 2] || (first && dp[i + 1][j]);
                } else {
                    dp[i][j] = first && dp[i + 1][j + 1];
                }
            }
        }
        return dp[0][0];
    }

    public static void main(String[] args) {
        System.out.println(isMatch("ray", "ra."));
        System.out.println(isMatch("raymond", "ra."));
        System.out.println(isMatch("chat", ".*at"));
        System.out.println(isMatch("chats", ".*at"));
    }
}

File: Day #1251/solution.rs
// Regex matching with '.' and '*' via DP; dp[i][j] = s[i:] matches p[j:]. O(m*n) time and space.
fn is_match(s: &str, p: &str) -> bool {
    let s: Vec<char> = s.chars().collect();
    let p: Vec<char> = p.chars().collect();
    let (m, n) = (s.len(), p.len());
    let mut dp = vec![vec![false; n + 1]; m + 1];
    dp[m][n] = true;
    for i in (0..=m).rev() {
        for j in (0..n).rev() {
            let first = i < m && (p[j] == s[i] || p[j] == '.');
            if j + 1 < n && p[j + 1] == '*' {
                dp[i][j] = dp[i][j + 2] || (first && dp[i + 1][j]);
            } else {
                dp[i][j] = first && dp[i + 1][j + 1];
            }
        }
    }
    dp[0][0]
}

fn main() {
    println!("{}", is_match("ray", "ra."));
    println!("{}", is_match("raymond", "ra."));
    println!("{}", is_match("chat", ".*at"));
    println!("{}", is_match("chats", ".*at"));
}
